Add arrow keys and asdf/hjkl keys to operate game

ghaerr opened the pull request:
Hello @FrenkelS,

Are you interested in this? At the price of swapping the ` key for ESC to pause game action, this PR adds up/down/left/right arrow keys to Doom, as well as allowing a/s/d/f and h/j/k/l for easier keyboard operation without using arrow keys. On my macOS laptop, it is much easier to play using hjkl/asdf or arrow keys than to use 2468.

Because the ANSI arrow key sequences contain [, the weapon drop is also changed to {, but in most cases ] (or }) accomplishes the same thing.

The game seems to play well with with ELKS configured for min memory in /bootopts. Using standard out-of-the-box configuration, it seems to run but I ran into a couple of Out of Memory errors, so memory must be tight.

On another note, the compelks.sh script should be mode 755 and a `set -e` added at the top, otherwise compilation errors don't stop the script and new version is silently not created.

FrenkelS commented:
Thanks for the pull request. The keyboard code / event_t code could use some improvements.
But this change breaks the cheat codes like IDDQD and IDKFA, because `d` is mapped to `KEYD_UP` while `C_Responder` in `m_cheat.c` still expects to receive `d`.

There are four places in the code that respond to keyboard events:
|File     |Responsibility|Function    |
|---------|--------------|------------|
|am_map.c |Auto map      |AM_Responder|
|g_game.c |Gameplay      |G_Responder |
|m_cheat.c|Cheats        |C_Responder |
|m_menu.c |Menu          |M_Responder |

I guess we could do the key mapping inside those `*_Responder` functions instead of in `I_StartTic`.

ghaerr commented:
How about if I just delete the "extra" a/s/d/f key handling? That would allow arrow keys to work and h/j/k/l as well?

FrenkelS commented:
I've added `set -e` to the top of `compelks.sh`. I'm not sure how to set mode 755 on the script here in Git. So I've added `chmod -v 755 compelks.sh` to `setenvwc.sh`.

FrenkelS commented:
I've added arrow keys support / Escape handling in `i_elks.c` and I've added support for asdf and hjkl to `g_game.c` and `m_menu.c`.

ghaerr commented:
> I've added arrow keys support / Escape handling in i_elks.c and I've added support for asdf and hjkl to g_game.c and m_menu.c.

Ok, that will help with the playability, thanks I'll close this now. I assume you've tested all of it.

> not sure how to set mode 755 on the script here in Git.
> chmod -v 755 compelks.sh to setenvwc.sh.

I don't recommend doing a chmod in setenvwc.sh, that might end up making git think the repo has changed every time setenvwc.sh is run; I think you can just `chmod 755 compelks.sh and then `git commit -a` and it should change it.

FrenkelS commented:
Whoa, I didn't know Git also supported access permissions.
